### PR TITLE
Apple Watch Complication

### PR DESCRIPTION
**Pull request by mzsanford**
This adds support for an Apple Watch complication as discussed in #119. It's a bit of a monster review but unfortunately adding a new embedded app is a bit of an all-or-nothing.

## Things to Note

1. The watch app requires a user to use the phone app. The watch app is getting data from the phone app via the complication updates API. Whenever the app loads new data or is closing it posts the latest known good state. To support watch-only installations the code base would need to be refactored heavily to make the data access into a shared framework. That felt beyond what I could do as a contributor.
2. I needed some smaller icons and tried to reuse the existing AppIcon where I could. For the smallest size (`miniCrab`) the  existing app icon was too busy. You only get 24x24 so I created a simpler icon but stayed with a turtle to match the main app. Earlier beta tests used a crab, which has since been removed.
3. This exposes the "half level" method also discussed in #119.
4. Some caode exists in the complication to display random characters on the watch (`templateForCharacter`). There is no path where this is called currently. That was  feature requested in #119 that is incomplete. I'm not sure if it should be finished or removed, honestly. Left here so we can decide.

## Future Ideas

I wanted to get this out as-is for the simplest version. If this is useful I'll be around to fix bugs that happen and finish/remove `templateForCharacter` as needed. After that I may attempt to refactor the complication into a shared component and two complications, "Tsurukame Reviews" and "Tsusukame Level" so user can have access to both.

### CODE REVIEW

**Comment by turtlemaster19** [comment]
@mzsanford I noticed some spaces in the file paths. It seems that this is undesirable if you look at other places in the repo. Perhaps instead of `Tsurukame Complication Extension`, we could just use `Complication` or `watch_complication`?

**Comment by mzsanford** [comment]
@turtlemaster19 good point. I renamed to remove the spaces … XCode makes that more complicated than seem reasonable so it took me some time to get to it.

File: ios/ComplicationExtension/ComplicationController.swift [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import ClockKit
+import os
+
+class ComplicationController: NSObject, CLKComplicationDataSource {
+  let TsurukameHighlightColor = UIColor.red
```
Comment (davidsansome) [inline ios/ComplicationExtension/ComplicationController.swift]: These constants should start with a lower-case letter, like `tsurukameHighlightColor`

File: ios/ComplicationExtension/ComplicationController.swift [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+      handler(entry)
+    } else {
+      handler(nil)
+    }
+  }
+
+  func getTimelineEntries(for _: CLKComplication, before _: Date, limit _: Int,
+                          withHandler handler: @escaping ([CLKComplicationTimelineEntry]?)
+                            -> Void) {
+    // Call the handler with the timeline entries prior to the given date
```
Comment (davidsansome) [inline ios/ComplicationExtension/ComplicationController.swift]: I was confused by this for a second.  Maybe reword this to be `// Don't provide any timeline entries in the past`

File: ios/ComplicationExtension/DataManager.swift [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+protocol DataManagerDelegate: AnyObject {
+  func onDataUpdated(data: UserData, dataSource: ComplicationDataSource)
+}
+
+enum ComplicationDataSource: Int {
+  case ReviewCounts, Level, Character
+}
+
+class DataManager {
+  private let UserDefaultsKeyData = "LastKnownData"
```
Comment (davidsansome) [inline ios/ComplicationExtension/DataManager.swift]: Same with these

File: ios/ComplicationExtension/DataManager.swift [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    return true
+  }
+
+  func addDelegate(_ delegate: DataManagerDelegate) {
+    delegates.append(delegate)
+  }
+
+  func removeDelegate(_ delegate: DataManagerDelegate) {
+    delegates.removeAll { (d) -> Bool in
```
Comment (davidsansome) [inline ios/ComplicationExtension/DataManager.swift]: I think this can be:

    delegates.removeAll { $0 === delegate }

File: ios/ComplicationExtension/SettingsController.swift [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    (.ReviewCounts, "Review Counts"),
+    (.Level, "Level"),
+    // Removed pending real data
+    // (.Character, "Kanji"),
+  ]
+
+  override func awake(withContext context: Any?) {
+    super.awake(withContext: context)
+
+    os_log("MZS - Awake with context: %{public}@", String(describing: context))
```
Comment (davidsansome) [inline ios/ComplicationExtension/SettingsController.swift]: Delete this line 😄

File: ios/ComplicationExtension/SettingsController.swift [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    for (idx, (dataSource, _)) in dataSourceOptions.enumerated() {
+      if dataSource == DataManager.sharedInstance.dataSource {
+        dataSourcePicker.setSelectedItemIndex(idx)
+      }
+    }
+  }
+
+  override func willActivate() {
+    // TODO: How to get context from ExtensionDelegate#handleUserActivity ?
+    os_log("MZS - will activate")
```
Comment (davidsansome) [inline ios/ComplicationExtension/SettingsController.swift]: This one too

File: ios/WatchHelper.swift [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+import WatchConnectivity
+#if os(watchOS)
+  import ClockKit
+#endif
+
+typealias ClientDelegateCallback = (([String: Any]) -> Void)
+typealias EpochTimeInt = Int64
+
+class WatchConnectionServerDelegate: NSObject, WCSessionDelegate {
+  override init() {
```
Comment (davidsansome) [inline ios/WatchHelper.swift]: I don't think you need this initialiser if all it's doing is calling the superclass's

File: ios/WatchHelper.swift [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+  #if os(iOS)
+    func sessionDidBecomeInactive(_: WCSession) {}
+
+    func sessionDidDeactivate(_: WCSession) {}
+  #endif
+}
+
+@objc class WatchHelper: NSObject {
+  public static let KeyReviewCount = "reviewCount"
```
Comment (davidsansome) [inline ios/WatchHelper.swift]: These too

File: ios/WatchHelper.swift [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+  let DuplicatePacketWindow = TimeInterval(30)
+
+  override init() {
+    super.init()
+    #if os(iOS)
+      startServerSession()
+    #endif
+  }
+
+  @objc static func sharedInstance() -> WatchHelper {
```
Comment (davidsansome) [inline ios/WatchHelper.swift]: Could this be a private(set) public(get) class variable?  I forget the exact syntax.

File: ios/ComplicationExtension/ComplicationController.swift (hunk 1 above)
Comment (mzsanford) [reply ios/ComplicationExtension/ComplicationController.swift]: Renamed all constants as suggested. Guess the language-hopping catches up with me sometimes :)

File: ios/ComplicationExtension/DataManager.swift (hunk 4 above)
Comment (mzsanford) [reply ios/ComplicationExtension/DataManager.swift]: Nice!

File: ios/ComplicationExtension/SettingsController.swift (hunk 6 above)
Comment (mzsanford) [reply ios/ComplicationExtension/SettingsController.swift]: Deleted this and a few others

File: ios/WatchHelper.swift (hunk 9 above)
Comment (mzsanford) [reply ios/WatchHelper.swift]: Changed to a `static let` which accomplishes that

**Comment by mzsanford** [comment]
@davidsansome Thanks for taking the time to review, I know it was a bit of a monster PR.
